src/sentry/digests/backends/redis.py

```diff
@@ -188,185 +188,205 @@ def add(self, key, record, increment_delay=None, maximum_delay=None):
-                # transaction as well.)
-                if not can_reschedule[True]:
-                    if len(items) == fetch_size:
-                        # There is nothing to reschedule in this chunk, but we
-                        # need check if there are others after this chunk.
-                        continue
-                    else:
-                        # There is nothing to unlock, and we've exhausted all items.
-                        break
+                with connection.pipeline() as pipeline:
+                    pipeline.multi()
 
-                try:
-                    with connection.pipeline() as pipeline:
-                        pipeline.multi()
+                    pipeline.zrem(
+                        make_schedule_key(self.namespace, SCHEDULE_STATE_WAITING),
+                        *[key for key, timestamp in items]
+                    )
 
-                        pipeline.zrem(
-                            make_schedule_key(self.namespace, SCHEDULE_STATE_READY),
-                            *[key for (lock, (key, timestamp)) in can_reschedule[True]]
-                        )
+                    pipeline.zadd(
+                        make_schedule_key(self.namespace, SCHEDULE_STATE_READY),
+                        *itertools.chain.from_iterable([(timestamp, key) for (key, timestamp) in items])
+                    )
 
-                        pipeline.zadd(
-                            make_schedule_key(self.namespace, SCHEDULE_STATE_WAITING),
-                            *itertools.chain.from_iterable([(timestamp, key) for (lock, (key, timestamp)) in can_reschedule[True]])
-                        )
+                    for key, timestamp in items:
+                        yield ScheduleEntry(key, timestamp)
 
-                        pipeline.execute()
-                finally:
-                    # Regardless of the outcome of the transaction, we should
-                    # try to unlock the items for processing.
-                    for lock, item in can_reschedule[True]:
-                        try:
-                            lock.release()
-                        except Exception as error:
-                            # XXX: This shouldn't be hit (the ``Lock`` code
-                            # should swallow the exception) but this is here
-                            # for safety anyway.
-                            logger.warning('Could not unlock %r: %s', item, error)
+                    pipeline.execute()
 
                 # If we retrieved less than the chunk size of items, we don't
                 # need try to retrieve more items.
-                if len(items) < fetch_size:
+                if len(items) < chunk:
                     break
             else:
                 raise RuntimeError('loop exceeded maximum iterations (%s)' % (maximum_iterations,))
 
+    def schedule(self, deadline, chunk=1000):
```

Comment:
This is the substantive part of the change, the rest was just moving lines around.